packages/backend/src/apps/m365-excel/common/workbook-session/index.ts

```diff
@@ -0,0 +1,306 @@
+ *      after some delay. This delay is set to an arbitrary estimate of the
+ *      longest time period Microsoft could possibly want to keep an inactive
+ *      session around for.
+ *
+ *    - That being said, Microsoft expires its sessions based on inactivity. To
+ *      prevent premature expiration, we optimistically extend the session ID
+ *      key's expiry each time we fetch it from redis.
+ *
+ * 4. Read locks should not be needed at all, as far as I can tell.
+ */
+export default class WorkbookSession {
+  private $: IGlobalVariable
+
+  private tenant: M365TenantInfo
+  private fileId: string
+  private sessionId: string
+
+  static async acquire(
+    $: IGlobalVariable,
+    fileId: string,
+  ): Promise<WorkbookSession> {
+    const tenant = getM365TenantInfo($.auth.data?.tenantKey as string)
+
+    let sessionId = await getSessionIdFromRedis(tenant, fileId)
+    if (!sessionId) {
+      sessionId = await refreshSessionId(tenant, fileId, $)
+    }
+
+    return new WorkbookSession($, tenant, fileId, sessionId)
+  }
+
+  private constructor(
+    $: IGlobalVariable,
+    tenant: M365TenantInfo,
+    fileId: string,
+    sessionId: string,
+  ) {
+    this.$ = $
+    this.fileId = fileId
+    this.tenant = tenant
+    this.sessionId = sessionId
+  }
+
+  private async _requestImpl<T>(
+    apiEndpoint: string,
+    method: NonNullable<AxiosRequestConfig['method']>,
+    config: AxiosRequestConfig | null,
+    canRetry: boolean,
+  ): Promise<AxiosResponse<T>> {
+    try {
+      const response = await this.$.http.request<T>({
+        ...(config ?? {}),
+        url: `/v1.0/sites/${this.tenant.sharePointSiteId}/drive/items/${this.fileId}/workbook${apiEndpoint}`,
+        method: method,
+        headers: {
+          ...(config?.headers ?? {}),
+          'workbook-session-id': this.sessionId,
+        },
+      })
+
+      // EDGE CASE
+      // ===
+      // If the session ID has changed after the request, the old session must
+      // have been invalidated by another worker due to a problem on
+      // Microsoft's servers. Since we just made a request to that invalid
+      // session, this means there is a chance of data corruption - so we throw
+      // an error here to inform the user about this.
+      //
+      // (This should occur very rarely, if at all...)
+      //
+      // Note that this doesn't save us from data loss; it's always possible for
+      // Microsoft's servers to return a successful response to us, then crash
+      // before propagating session data to other machines. This error is just a
+      // nice-to-have.
+
+      const sessionIdInRedis = await getSessionIdFromRedis(
+        this.tenant,
+        this.fileId,
+      )
+      if (sessionIdInRedis !== this.sessionId) {
+        throw new Error(
+          "M365 servers encountered an error and invalidated the current Excel session; please double check the file's data",
+        )
```

Comment:
other reasons for this error could be redis related (e.g. failed to write to redis, failed to read from redis, data was deleted..)

Reply:
Have word-smithed the error message a little bit